# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/DOMTypes.ipdlh
# Commit: 0d0f19a4db70
# Full Hash: 0d0f19a4db70e734724b99c959148fddef94811e
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2022-08-03 03:42:48
# Regressor Bug: 1781129
# File Overlap Count: 1
# Description:
#   Bug 1781129 - Part 1: Use BigBuffer for IPCDataTransfer, r=edgar
#   
#   The IPCDataTransfer type is used to transfer Clipboard/Drag & Drop payloads
#   over IPC to allow them to be written to or read from the relevant system
#   interfaces. Previously, the system which was used was somewhat complex, and
# ==============================================================================

diff -r 65a0b02c1909 -r 0d0f19a4db70 dom/ipc/DOMTypes.ipdlh
--- a/dom/ipc/DOMTypes.ipdlh	Tue Aug 02 21:34:06 2022 +0300
+++ b/dom/ipc/DOMTypes.ipdlh	Tue Aug 02 18:09:40 2022 +0000
@@ -42,7 +42,6 @@
 using nsSizeMode from "nsIWidgetListener.h";
 using ScrollbarPreference from "mozilla/ScrollbarPreferences.h";
 using mozilla::gfx::SurfaceFormat from "mozilla/gfx/Types.h";
-using TransferableDataType from "ipc/nsTransferableIPC.h";
 [RefCounted] using class nsIPrincipal from "nsIPrincipal.h";
 using mozilla::dom::MaybeDiscardedBrowsingContext from "mozilla/dom/BrowsingContext.h";
 [RefCounted] using class nsIURI from "nsIURI.h";
@@ -53,6 +52,7 @@
 using mozilla::TimeStamp from "mozilla/TimeStamp.h";
 [RefCounted] using class mozilla::dom::BrowsingContext from "mozilla/dom/BrowsingContext.h";
 [RefCounted] using class mozilla::RemoteLazyInputStream from "mozilla/RemoteLazyInputStream.h";
+[MoveOnly] using class mozilla::ipc::BigBuffer from "mozilla/ipc/BigBuffer.h";
 
 namespace mozilla {
 namespace dom {
@@ -101,28 +101,55 @@
   MessageDataType data;
 };
 
-union IPCDataTransferData
+struct IPCDataTransferString
 {
-  nsString;  // text
-  Shmem;     // images using Shmem
-  IPCBlob;   // files
+  BigBuffer data;
+};
+
+struct IPCDataTransferCString
+{
+  BigBuffer data;
 };
 
-struct IPCDataTransferImage
+struct IPCDataTransferInputStream
 {
+  // NOTE: Editor currently relies on these input streams being synchronous, so
+  // we can't safely serialize them using IPCStream (see bug 1778565). Instead,
+  // they're serialized as a `BigBuffer`, and converted to a nsStringInputStream
+  // on the receiving side. If we are able to use async streams reliably in the
+  // future, we could consider switching the code which adds `nsIInputStream`s
+  // to the transferable to use `BlobImpl` instead, for more consistency between
+  // image formats.
+  BigBuffer data;
+};
+
+struct IPCDataTransferImageContainer
+{
+  BigBuffer data;
   uint32_t width;
   uint32_t height;
   uint32_t stride;
   SurfaceFormat format;
 };
 
+struct IPCDataTransferBlob
+{
+  IPCBlob blob;
+};
+
+union IPCDataTransferData
+{
+  IPCDataTransferString;
+  IPCDataTransferCString;
+  IPCDataTransferInputStream;
+  IPCDataTransferImageContainer;
+  IPCDataTransferBlob;
+};
+
 struct IPCDataTransferItem
 {
   nsCString flavor;
   IPCDataTransferData data;
-  TransferableDataType dataType;
-  // The image details are only used when transferring images.
-  IPCDataTransferImage? imageDetails;
 };
 
 struct IPCDataTransfer